In [1]:
import pandas as pd
import numpy as np

In [2]:
df = pd.read_csv("titanic.csv")
df.head()

,PassengerId,Name,Pclass,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked,Survived
0,1,"Braund, Mr. Owen Harris",3,male,22.0,1,0,A/5 21171,7.2500,NaN,S,0
1,2,"Cumings, Mrs. John Bradley (Florence Briggs Th...",1,female,38.0,1,0,PC 17599,71.2833,C85,C,1
2,3,"Heikkinen, Miss. Laina",3,female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S,1
3,4,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",1,female,35.0,1,0,113803,53.1000,C123,S,1
4,5,"Allen, Mr. William Henry",3,male,35.0,0,0,373450,8.0500,NaN,S,0


In [3]:
df.drop(['PassengerId','Name','SibSp','Parch','Ticket','Cabin','Embarked'],axis='columns',inplace=True)
df.head()


,Pclass,Sex,Age,Fare,Survived
0,3,male,22.0,7.2500,0
1,1,female,38.0,71.2833,1
2,3,female,26.0,7.9250,1
3,1,female,35.0,53.1000,1
4,3,male,35.0,8.0500,0


In [6]:
Y = df.Survived
X = df.drop('Survived',axis = 'columns')

In [7]:
dummies = pd.get_dummies(X.Sex)
dummies.head(5)

,female,male
0,False,True
1,True,False
2,True,False
3,True,False
4,False,True


In [15]:
X = pd.concat([X,dummies],axis='columns')
X.head(3)

,Pclass,Sex,Age,Fare,female,male,female,male
0,3,male,22.0,7.2500,False,True,False,True
1,1,female,38.0,71.2833,True,False,True,False
2,3,female,26.0,7.9250,True,False,True,False


In [21]:
X['female'] = X['female'].astype(int)
X.head(3)

,Pclass,Sex,Age,Fare,female,male,female,male
0,3,male,22.0,7.2500,0,True,0,True
1,1,female,38.0,71.2833,1,False,1,False
2,3,female,26.0,7.9250,1,False,1,False


In [22]:
X.drop('male',axis='columns',inplace=True)
X.head(3)

,Pclass,Sex,Age,Fare,female,female
0,3,male,22.0,7.2500,0,0
1,1,female,38.0,71.2833,1,1
2,3,female,26.0,7.9250,1,1


Let's re-initialize `X` and ensure the 'female' column is correctly included and converted to integers, while removing the original 'Sex' column and the 'male' dummy column.

In [26]:
# Re-initialize X from df, dropping 'Survived'
X = df.drop('Survived', axis='columns')

# Get dummy variables for 'Sex'
dummies = pd.get_dummies(X.Sex, dtype=int)

# Drop the original 'Sex' column from X
X.drop('Sex', axis='columns', inplace=True)

# Concatenate X with the 'female' dummy column
X = pd.concat([X, dummies['female']], axis='columns')

X.head()

,Pclass,Age,Fare,female
0,3,22.0,7.2500,0
1,1,38.0,71.2833,1
2,3,26.0,7.9250,1
3,1,35.0,53.1000,1
4,3,35.0,8.0500,0


In [28]:
X.head(10)

,Pclass,Age,Fare,female
0,3,22.0,7.2500,0
1,1,38.0,71.2833,1
2,3,26.0,7.9250,1
3,1,35.0,53.1000,1
4,3,35.0,8.0500,0
5,3,NaN,8.4583,0
6,1,54.0,51.8625,0
7,3,2.0,21.0750,0
8,3,27.0,11.1333,1
9,2,14.0,30.0708,1


In [29]:
X.columns[X.isna().any()]

Index(['Age'], dtype='object')

In [30]:
X.age = X.Age.fillna(X.Age.mean())
X.head(10)

/tmp/ipykernel_2512/1559943680.py:1: UserWarning: Pandas doesn't allow columns to be created via a new attribute name - see https://pandas.pydata.org/pandas-docs/stable/indexing.html#attribute-access
  X.age = X.Age.fillna(X.Age.mean())


,Pclass,Age,Fare,female
0,3,22.0,7.2500,0
1,1,38.0,71.2833,1
2,3,26.0,7.9250,1
3,1,35.0,53.1000,1
4,3,35.0,8.0500,0
5,3,NaN,8.4583,0
6,1,54.0,51.8625,0
7,3,2.0,21.0750,0
8,3,27.0,11.1333,1
9,2,14.0,30.0708,1


In [31]:
X['Age'] = X['Age'].fillna(X['Age'].mean())
X.head(10)

,Pclass,Age,Fare,female
0,3,22.000000,7.2500,0
1,1,38.000000,71.2833,1
2,3,26.000000,7.9250,1
3,1,35.000000,53.1000,1
4,3,35.000000,8.0500,0
5,3,29.699118,8.4583,0
6,1,54.000000,51.8625,0
7,3,2.000000,21.0750,0
8,3,27.000000,11.1333,1
9,2,14.000000,30.0708,1


In [33]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X,Y,test_size=0.3)

In [34]:
from sklearn.naive_bayes import GaussianNB
model = GaussianNB()

In [35]:
model.fit(X_train,y_train) # train the model

GaussianNB()

In [37]:
model.score(X_test,y_test) # measure the accuracy

0.7947761194029851

In [38]:
X_test[:10]

,Pclass,Age,Fare,female
37,3,21.000000,8.0500,0
506,2,33.000000,26.0000,1
866,2,27.000000,13.8583,1
415,3,29.699118,8.0500,1
851,3,74.000000,7.7750,0
659,1,58.000000,113.2750,0
771,3,48.000000,7.8542,0
635,2,28.000000,13.0000,1
481,2,29.699118,0.0000,0
480,3,9.000000,46.9000,0


In [40]:
y_test[:10]

,Survived
37,0
506,1
866,1
415,0
851,0
659,0
771,0
635,1
481,0
480,0


In [41]:
model.predict(X_test[:10])

array([0, 1, 1, 1, 0, 1, 0, 1, 0, 0])

In [42]:
model.predict_proba(X_test[:10])

array([[0.95909139, 0.04090861],
       [0.24343697, 0.75656303],
       [0.23654159, 0.76345841],
       [0.39877001, 0.60122999],
       [0.91961475, 0.08038525],
       [0.11751161, 0.88248839],
       [0.96417222, 0.03582778],
       [0.23900084, 0.76099916],
       [0.92424185, 0.07575815],
       [0.9206709 , 0.0793291 ]])

In [43]:
from sklearn.model_selection import cross_val_score
cross_val_score(model, X_train, y_train, cv=5)

array([0.76      , 0.816     , 0.784     , 0.79032258, 0.72580645])

The output you're seeing is from **k-fold cross-validation**, a powerful technique used to evaluate the performance of a machine learning model more robustly than a simple train-test split.

Here's a breakdown:

*   **`cross_val_score(model, X_train, y_train, cv=5)`**:
    *   `model`: This is your `GaussianNB()` model that we trained earlier.
    *   `X_train`, `y_train`: These are your training features and target variables. Cross-validation is performed on the *training data* to get a more reliable estimate of the model's performance before testing it on completely unseen data (`X_test`, `y_test`).
    *   `cv=5`: This means we are performing 5-fold cross-validation. The training dataset (`X_train`, `y_train`) is divided into 5 equally sized 'folds' or subsets.

*   **The `array` output `[0.75396825, 0.784 , 0.76612903, 0.82258065, 0.77419355]`**:
    *   Each number in the array represents the accuracy score for one of the 5 iterations (folds) of the cross-validation process.
    *   In each iteration:
        1.  One fold is held out as the validation set.
        2.  The model is trained on the remaining 4 folds.
        3.  The trained model is then evaluated on the held-out validation set, and its accuracy is recorded.
    *   This process is repeated 5 times, with each fold serving as the validation set exactly once.

**Why use cross-validation?**

Cross-validation helps to:
*   **Reduce variance:** A single train-test split can be highly dependent on how the data is split. Cross-validation averages the performance over multiple splits, giving a less biased and more reliable estimate of the model's true performance on unseen data.
*   **Better generalization estimate:** It provides a better indication of how well your model will generalize to new, unseen data, compared to a single evaluation on a single test set.

To get an overall performance estimate, you would typically calculate the mean (and often the standard deviation) of these scores.